tests/integration_test/http_method_override_test.go

```diff
@@ -0,0 +1,260 @@
+				Get: "/echoMethod",
+			},
+		},
+		{
+			Selector: "1.echo_api_endpoints_cloudesf_testing_cloud_goog.echoPOST",
+			Pattern: &annotationspb.HttpRule_Post{
+				Post: "/echoMethod",
+			},
+		},
+	})
+
+	defer s.TearDown()
+	if err := s.Setup(utils.CommonArgs()); err != nil {
+		t.Fatalf("fail to setup test env, %v", err)
+	}
+
+	testData := []struct {
+		desc          string
+		path          string
+		method        string
+		headers       map[string]string
+		wantResp      string
+		httpCallError error
+	}{
+		{
+			desc:   "Overridden POST is received as GET",
+			path:   "/echoMethod?key=api-key",
+			method: "POST",
+			headers: map[string]string{
+				"X-HTTP-Method-Override": "GET",
+			},
+			wantResp: `{"RequestMethod":"GET"}`,
+		},
+		{
+			desc:   "Overridden DELETE is received as POST",
+			path:   "/echoMethod?key=api-key",
+			method: "DELETE",
+			headers: map[string]string{
+				"X-HTTP-Method-Override": "POST",
+			},
+			wantResp: `{"RequestMethod":"POST"}`,
+		},
+		{
+			desc:   "Overridden POST is rejected because DELETE is not defined in the service config",
+			path:   "/echoMethod?key=api-key",
+			method: "POST",
+			headers: map[string]string{
+				"X-HTTP-Method-Override": "DELETE",
+			},
+			httpCallError: fmt.Errorf("404 Not Found"),
+		},
+	}
+	for _, tc := range testData {
+		url := fmt.Sprintf("http://localhost:%v%v", s.Ports().ListenerPort, tc.path)
+		gotResp, err := client.DoWithHeaders(url, tc.method, "test-body", tc.headers)
+
+		if tc.httpCallError == nil {
+			if err != nil {
+				t.Fatal(err)
+			}
+		} else {
+			if !strings.Contains(err.Error(), tc.httpCallError.Error()) {
+				t.Errorf("expected Http call error: %v, got: %v", tc.httpCallError, err)
+			}
+			continue
+		}
+		gotRespStr := string(gotResp)
+		if err := util.JsonEqual(tc.wantResp, gotRespStr); err != nil {
+			t.Errorf("Test(%s) fails: \n %s", tc.desc, err)
+		}
+	}
+}
+
+func TestMethodOverrideBackendBody(t *testing.T) {
+	t.Parallel()
+
+	s := env.NewTestEnv(comp.TestMethodOverrideBackendBody, platform.EchoSidecar)
+	defer s.TearDown()
+	if err := s.Setup(utils.CommonArgs()); err != nil {
+		t.Fatalf("fail to setup test env, %v", err)
+	}
+
+	testData := []struct {
+		desc          string
+		path          string
+		method        string
+		headers       map[string]string
+		body          string
+		wantResp      string
+		httpCallError error
+	}{
+		{
+			desc:   "Overridden PUT has body",
+			path:   "/echo?key=api-key",
+			method: "PUT",
+			headers: map[string]string{
+				"X-HTTP-Method-Override": "POST",
+			},
+			body:     `hello`,
+			wantResp: `{"message":"hello"}`,
+		},
+		{
+			desc:   "Overridden GET has no body for the backend to handle",
+			path:   "/echo?key=api-key",
+			method: "GET",
+			headers: map[string]string{
+				"X-HTTP-Method-Override": "POST",
+			},
+			body:          "This body will not be sent in the request",
+			httpCallError: fmt.Errorf(`{"code":500,"message":"Could not get body: EOF"}`),
```

Comment:
what will happen for POST overrided by GET?  Will the body be send to the backend,  will the backend echo it body?

Reply:
Yes, the backend will receive the body. This is spec complaint. Added a test.